typing/typemod.ml

```diff
@@ -85,14 +87,68 @@ let extract_sig_open env loc mty =
 
 (* Compute the environment after opening a module *)
 
-let type_open_ ?used_slot ?toplevel ovf env loc lid =
-  let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
-  match Env.open_signature ~loc ?used_slot ?toplevel ovf path env with
-  | Some env -> path, env
-  | None ->
-      let md = Env.find_module path env in
-      ignore (extract_sig_open env lid.loc md.md_type);
-      assert false
+let type_module_fwd : (Env.t -> Parsetree.module_expr ->
+                       Typedtree.module_expr) ref =
+  ref (fun _ _ -> assert false)
+
+let mod_ident_counter = ref 0
+
+let gen_mod_ident () =
+  let n = !mod_ident_counter in
+  incr mod_ident_counter;
+  let ident = Ident.create (Printf.sprintf "M#%d" n) in
+  ident
+
+let open_struct_level = ref 0
+let in_nested_struct () = !open_struct_level <> 0
+let enter_struct () = incr open_struct_level
+let leave_struct () = decr open_struct_level
+
+let open_generated s = String.contains s '#'
+
+let type_open_ ?used_slot ?toplevel ovf env loc me =
+  match me.pmod_desc with
+  | Pmod_functor _ | Pmod_extension _ ->
+      raise(Error(me.pmod_loc, env, Invalid_open me))
+  | Pmod_ident lid ->
+      let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
+      begin
+      match Env.open_signature ~loc ?used_slot ?toplevel ovf path env with
+      | Some env ->
+          let tme =
+            {
+              mod_desc=Tmod_ident (path, lid);
+              mod_loc=lid.loc;
+              mod_type=Mty_ident path;
+              mod_env=env;
+              mod_attributes=me.pmod_attributes
+            } in
+          None, tme, env
+      | None ->
+          (* use [extract_sig_open] to provide better error message *)
+          let md = Env.find_module path env in
+          ignore (extract_sig_open env lid.loc md.md_type);
+          assert false
+      end
+  | Pmod_structure _ | Pmod_apply _ | Pmod_constraint _ | Pmod_unpack _ ->
+      enter_struct ();
+      let ident = gen_mod_ident () in
+      let tme = !type_module_fwd env me in
+      leave_struct ();
+      (match tme.mod_type with
+       | Mty_signature _ | Mty_ident _ -> ()
+       | Mty_functor _ -> raise(Error(me.pmod_loc, env, Invalid_open me))
+       | Mty_alias _ -> assert false);
+      let md = {
+        md_type = tme.mod_type;
+        md_loc = me.pmod_loc;
+        md_attributes = me.pmod_attributes;
+      } in
+      let newenv = Env.enter_module_declaration ident md env in
+      let root = Pident ident in
+      match Env.open_signature ~loc ?used_slot ?toplevel ovf root newenv with
+      | None -> assert false (* not possible to illegal module_expr *)
```

Comment:
This is where I would raise Invalid_open.

Reply:
addressed via https://github.com/ocaml/ocaml/pull/1506#discussion_r178453010